portfolio/src/main/java/com/google/sps/servlets/ThumbsUpServlet.java

```diff
@@ -31,54 +31,140 @@
+
         // Using the id, get the comment's key.
         Key commentEntityKey;
         try {
             commentEntityKey = KeyFactory.createKey("Comment", id);
         } catch(Exception e) {
-            response.setContentType("text/html;");
-            response.getWriter().println("Unable to get comment's key.");
-            return;
+            return null;
         }
 
-        // Instantiate datastore.
-        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-
         // Get the comment entity using its key.
         Entity commentEntity;
         try {
             commentEntity = datastore.get(commentEntityKey);
         } catch(Exception e) {
-            response.setContentType("text/html;");
-            response.getWriter().println("Unable to get comment.");
-            return;
+            return null;
         }
 
+        return commentEntity;
+    }
+
+    private void changePopularity(Entity commentEntity, long value) {
+        // Instantiate datastore
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
         // Get the previous thumbs up value.
         long prevThumbsUp = (long) commentEntity.getProperty("thumbsup");
         // Get the previous popularity value.
         long prevPopularity = (long) commentEntity.getProperty("popularity");
-        long newThumbsUp = prevThumbsUp + 1;
-        long newPopularity = prevPopularity + 1;
+
+        long newThumbsUp = prevThumbsUp + value;
+        long newPopularity = prevPopularity + value;
         
-        // Update the thumbs up property to be the previous value plus one
+        // Update the thumbs up property to be the previous value plus one.
         commentEntity.setProperty("thumbsup", newThumbsUp);
         // Update the popularity property to be the previous one plus one.
         commentEntity.setProperty("popularity", newPopularity);
 
-        // Add the updated entity back in the datastore.
+        // Add the updated entity back in the datastore
         datastore.put(commentEntity);
+    }
 
-        response.sendRedirect("/contact.html");
-        return;
+    // Accesses the datastore to get the UserInfo entity. Returns the entity or null if one does not exist.
+    private Entity getUserInfoEntity() {
+        UserService userService = UserServiceFactory.getUserService();
+        String id = userService.getCurrentUser().getUserId();
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Filter queryFilter = new FilterPredicate("id", Query.FilterOperator.EQUAL, id);
+        Query query = new Query("UserInfo").setFilter(queryFilter);
+        PreparedQuery results = datastore.prepare(query); 
+        Entity userInfoEntity = results.asSingleEntity(); 
+
+        return userInfoEntity;
+    }
+
+    private void addToLikedComments(Entity userInfoEntity, Entity commentEntity) {
```

Comment:
similar to unliked comments, these two methods need to be made thread-safe using transactions